File: 1423-maximum-points-you-can-obtain-from-cards/1423-maximum-points-you-can-obtain-from-cards.py
class Solution:
    def maxScore(self, cardPoints: List[int], k: int) -> int:
        total_sum=0
        n=len(cardPoints)
        for c in cardPoints:
            total_sum+=c
        prefix=[0]*(n+1)
        prefix[0]=cardPoints[0]
        for i in range(1,n+1):
            prefix[i]=prefix[i-1]+cardPoints[i-1]
        #print(prefix)    
        maxi=float("-inf")    
        for i in range(k+1):
            maxi=max(maxi,total_sum-(prefix[i+(n-k)]-prefix[i]))
        return maxi    
            
            
        

File: 1423-maximum-points-you-can-obtain-from-cards/1423-maximum-points-you-can-obtain-from-cards.cpp
class Solution {
public:
    int maxScore(vector<int>& cardPoints, int k) {      
        //Instead of finding the maximum score we can find the consective minimum score (subarray) using the concept of prefix sum
        //answer will be total_sum-mini;
        int mini=INT_MIN;
        int n=cardPoints.size();
        vector<int> prefix(n+1,0);
        for(int i=0;i<n;i++){
            prefix[i+1]=prefix[i]+cardPoints[i];
        }
        int total=prefix[n];
        for(int i=0;i<=k;i++){
            mini=max(mini,total-(prefix[i+n-k]-prefix[i]));
            //mini=min(mini,prefix[i+n-k]-prefix[i])
        }
        //return total-mini;
        return mini;
        
    }
};